vm/src/slots.rs

```diff
@@ -530,6 +531,50 @@ pub trait AsBuffer: PyValue {
     fn as_buffer(zelf: &PyRef<Self>, vm: &VirtualMachine) -> PyResult<PyBuffer>;
 }
 
+#[pyimpl]
+pub trait AsMapping: PyValue {
+    #[pyslot]
+    fn tp_as_mapping(zelf: &PyObjectRef, vm: &VirtualMachine) -> PyResult<PyMapping> {
+        let zelf = zelf
+            .downcast_ref()
+            .ok_or_else(|| vm.new_type_error("unexpected payload for as_mapping".to_owned()))?;
+        Self::as_mapping(zelf, vm)
+    }
```

Comment:
`as_mapping` looks like not requiring any argument. Maybe this will be ok.
```suggestion
    #[pyslot]
    fn as_mapping() -> PyMapping {
            PyMapping {
                length: Some(Self::length),
                subscript: Some(Self::subscript),
                ass_subscript: Some(Self::ass_subscript),
            }
    }
```

Reply:
I cannot declare empty parameter function as slot according to below error message 😢 I think function signature for pyslot must be a `fn name(&PyObjectRef, &VirtualMachine)`
```rust
non-primitive cast: `fn() -> std::result::Result<builtins::dict::PyMapping, 
pyobjectrc::PyRef<exceptions::types::PyBaseException>> {<Self as slots::AsMapping>::slot_as_mapping}` as `for<'r, 's> 
fn(&'r pyobjectrc::PyObjectRef, &'s vm::VirtualMachine) -> std::result::Result<builtins::dict::PyMapping, 
pyobjectrc::PyRef<exceptions::types::PyBaseException>>`
```

Although this is little bit dirty, how about this one?
```rust
#[pyslot]
fn slot_as_mapping(_zelf: &PyObjectRef, _vm: &VirtualMachine) -> PyResult<PyMapping> {
    Self::as_mapping()
}
fn as_mapping() -> PyResult<PyMapping>;
```